packages/embarkjs/snark/src/circuit.ts

```diff
@@ -0,0 +1,97 @@
+import { CircuitSetup } from ".";
+import * as snarkjs from "snarkjs";
+const { unstringifyBigInts } = require("snarkjs/src/stringifybigint");
+
+const LOG_PREFIX = "[embarkjs-snark]: ";
+
+export default class Circuit {
+  constructor(private setup: CircuitSetup) {
+    this.setup.provingKey = unstringifyBigInts(this.setup.provingKey);
+    this.setup.verificationKey = unstringifyBigInts(this.setup.verificationKey);
+  }
+
+  private generateSolidityInputs(publicSignals, proof): string[] {
+    publicSignals = unstringifyBigInts(publicSignals);
+    proof = unstringifyBigInts(proof);
+
+    const p256 = (n) => {
+      let nstr = n.toString(16);
+      while (nstr.length < 64) { nstr = "0" + nstr; }
+      nstr = `0x${nstr}`;
+      return nstr;
+    };
+
+    let inputs = "";
+    for (const publicSignal of publicSignals) {
+      if (inputs !== "") { inputs = inputs + ","; }
+      inputs = inputs + p256(publicSignal);
+    }
+
+    let S;
+    if ((typeof proof.protocol === "undefined") || (proof.protocol === "original")) {
+      S = [
+        [proof.pi_a[0], proof.pi_a[1]],
+        [proof.pi_ap[0], proof.pi_ap[1]],
+        [[proof.pi_b[0][1], proof.pi_b[0][0]], [proof.pi_b[1][1], proof.pi_b[1][0]]],
+        [proof.pi_bp[0], proof.pi_bp[1]],
+        [proof.pi_c[0], proof.pi_c[1]],
+        [proof.pi_cp[0], proof.pi_cp[1]],
+        [proof.pi_h[0], proof.pi_h[1]],
+        [proof.pi_kp[0], proof.pi_kp[1]]
+      ];
+    } else if ((proof.protocol === "groth") || (proof.protocol === "kimleeoh")) {
+      S = [
+        [proof.pi_a[0], proof.pi_a[1]],
+        [[proof.pi_b[0][1], proof.pi_b[0][0]], [proof.pi_b[1][1], proof.pi_b[1][0]]],
+        [proof.pi_c[0], proof.pi_c[1]]
```

Comment:
I added a comment to the function, hopefully it can help. To be honest, I'm not 100% sure how the mathematics behind it work, but this function takes certain inputs from the proof, pads them to 64 bytes, and arranges them in an array that can be used to call the `verifyProof` function of the verification contract that was generated using the trusted setup. I took what was in snarkjs cli only (this function isn't extracted and exported in the source unfortunately), and tried to make it easier to read haha. 

If we ever need to work on this, we probably need to speak to the iden3 guys 😂